zerver/lib/test_classes.py

```diff
@@ -165,21 +166,31 @@ def validate_api_response_openapi(self, url: str, method: str, result: HttpRespo
             content = ujson.loads(result.content)
         except ValueError:
             return
-        url = re.sub(r"\?.*", "", url)
-        validate_against_openapi_schema(content,
-                                        url.replace("/json/", "/").replace("/api/v1/", "/"),
-                                        method, str(result.status_code))
+        json_url = False
+        if url.startswith('/json'):
+            json_url = True
+        url_split = url.split('?')
+        # In some cases the query parameters are defined in the url itself. In such cases
+        # The `data` argument of our function is not used. Hence get `data` arguement
+        # from url.
+        if len(url_split) == 2:
+            data = urllib.parse.parse_qs(url_split[1])
+        url = url_split[0]
+        url = url.replace("/json/", "/").replace("/api/v1/", "/")
```

Comment:
Let's extract a separate function for this parsing logic, and have it be

`relative_url = extract_api_suffix_url(url)`